### PR TITLE
Fixes #33484 - add setting to opt-in/out the new host details page

### PR DESCRIPTION
**Pull request by amirfefer**
This setting opt-in/out the new host details page. 
When it's off,  any link to `/hosts/:id` redirects to the old host show page

### CODE REVIEW

**Comment by ezr-ondrej** [comment]
Could we have a button for switching to the new page on the old host details page?
We would cherry-pick that to 3.0 to get more exposure for the new page and more feedback.

**Comment by ekohl** [comment]
Linking back: https://community.theforeman.org/t/foreman-3-0-new-host-detail-page-feedback/25281

File: app/models/setting/general.rb [hunk 1]
```diff
@@ -35,6 +35,7 @@ def self.default_settings
       set('default_locale', N_("Language to use for new users"), nil, N_('Default language'), nil, { :collection => proc { locales } }),
       set('default_timezone', N_("Timezone to use for new users"), nil, N_('Default timezone'), nil, { :collection => proc { timezones } }),
       set('instance_title', N_("The instance title is shown on the top navigation bar (requires reload of page)."), nil, N_('Instance title')),
+      set('host_details_ui', N_("Foreman will load the new ui for host details"), true, N_('New host details UI')),
```
Comment (ezr-ondrej) [inline app/models/setting/general.rb]:
```suggestion
      set('host_details_ui', N_("Foreman will load the new UI for host details"), false, N_('New host details UI')),
```

**Comment by amirfefer** [comment]
> Could we have a button for switching to the new page on the old host details page?

@ezr-ondrej, it's not trivial, when the `host_details_ui` setting is off,  navigating to `/hosts/<id>` redirects automatically to `/hosts/legacy/<id>`

File: webpack/assets/javascripts/react_app/components/HostDetails/index.js [hunk 2]
```diff
@@ -1,155 +1,46 @@
...
+    case RENDER_LEGACY:
+      return <Redirect to={`${LEGACY_DETAILS_PATH}/${id}`} />;
+    case RENDER_REACT:
+      return <HostDetailsPage {...props} />;
+    default:
+      return hostDetailsUI ? (
+        <HostDetailsPage {...props} />
+      ) : (
+        <Redirect to={`${LEGACY_DETAILS_PATH}/${id}`} />
+      );
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/HostDetails/index.js]: @ezr-ondrej, this allows consumers to switch programmatically between react / legacy page.
switchers buttons will be added in a separate PR

**Comment by amirfefer** [comment]
I've blocked this PR in favor of these [enhancements](https://projects.theforeman.org/issues/33499), so it would be easier to cp into 3.01

File: webpack/assets/javascripts/react_app/components/HostDetails/HostDetailsPage.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
+/* eslint-disable camelcase */
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/HostDetails/HostDetailsPage.js:1]: this file contains the same content, just moved from `index.js` -> `HostDetailsPage.js `

**Comment by amirfefer** [comment]
@ezr-ondrej, 
fixed and rebased :)

File: config/routes.rb [hunk 4]
```diff
@@ -545,7 +545,8 @@
 
   match 'host_statuses' => 'react#index', :via => :get
   constraints(id: /[^\/]+/) do
-    match 'experimental/hosts/:id' => 'react#index', :via => :get, :as => :host_details_page
+    get 'legacy/hosts/:id' => 'hosts#show'
+    match 'hosts/:id' => 'react#index', :via => :get, :as => :host_details_page
```
Comment (ezr-ondrej) [inline config/routes.rb]: The tests are failing on this, could we flip to something like:
```suggestion
    get 'hosts/:id' => 'hosts#show'
    match 'fancy/hosts/:id' => 'react#index', :via => :get, :as => :host_details_page
```

**Comment by amirfefer** [comment]
[test foreman]

**Comment by amirfefer** [comment]
[katello test]

File: config/routes.rb (hunk 4 above)
Comment (amirfefer) [reply config/routes.rb]: how about `/old/hosts/:id`?

Comment (ezr-ondrej) [reply config/routes.rb]: I'm saying that change of the URL of the old page make the tests to fail as they expect the url to be `/hosts/:id` that can be fixed in tests tho if you want.
But going with the new UI being on `/hosts/:id` makes it feel like its already default UI :)

File: webpack/assets/javascripts/react_app/routes/ForemanSwitcher/index.js [hunk 5]
```diff
@@ -5,12 +5,16 @@ import { Switch, Route } from 'react-router-dom';
...
 import { fallbackRoute } from '../RoutingService';
 import { selectRoutes } from '../RouterSelector';
+import { OVERRIDE_ROUTES } from './OverrideRoutes';
 
 const ForemanSwitcher = ({ children: coreRoutes }) => {
   const routes = useSelector(() => selectRoutes(coreRoutes), shallowEqual);
 
   return (
     <Switch>
+      {OVERRIDE_ROUTES.map(({ path, exact }) => (
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/routes/ForemanSwitcher/index.js:15]: This `OVERRIDE_ROUTES` array contains routes that overlap with the server-side routes.
i.e. `hosts/:id` (client)  <-> `hosts/new` (server)

File: app/assets/javascripts/host_edit.js [hunk 6]
```diff
@@ -164,7 +164,6 @@ function submit_with_all_params() {
   $('body').css('cursor', 'progress');
   clear_errors();
   animate_progress();
-
```
Comment (ezr-ondrej) [inline app/assets/javascripts/host_edit.js:167]: ? this is weird to edit file that doesn't need to be edited.

File: app/controllers/hosts_controller.rb [hunk 7]
```diff
@@ -60,7 +60,7 @@ def index(title = nil)
     end
   end
 
-  def show
+  def legacy_show
```
Comment (ezr-ondrej) [inline app/controllers/hosts_controller.rb:63]: I'll not block on this, but I'd really feel better if we have rather `show` as the old page and `react_show` as the new one.

File: app/controllers/hosts_controller.rb [hunk 8]
```diff
@@ -24,7 +24,7 @@ class HostsController < ApplicationController
                         select_multiple_power_state update_multiple_power_state)
 
   before_action :ajax_request, :only => AJAX_REQUESTS
-  before_action :find_resource, :only => [:show, :clone, :edit, :update, :destroy, :review_before_build,
+  before_action :find_resource, :only => [:show, :legacy_show, :clone, :edit, :update, :destroy, :review_before_build,
```
Comment (ezr-ondrej) [inline app/controllers/hosts_controller.rb:27]: I do not think that we need to load the resource for the react page, do we?

File: app/helpers/host_description_helper.rb [hunk 9]
```diff
@@ -136,7 +136,7 @@ def base_host_overview_buttons(host)
       { :button => link_to_if_authorized(_("Audits"), { :controller => 'audits', :action => 'index', :search => "host = #{host.name}" }, :title => _("Host audit entries"), :class => 'btn btn-default'), :priority => 100 },
       ({ :button => link_to_if_authorized(_("Facts"), hash_for_host_facts_path(:host_id => host), :title => _("Browse host facts"), :class => 'btn btn-default'), :priority => 200 } if host.fact_values.any?),
       ({ :button => link_to_if_authorized(_("Reports"), hash_for_host_config_reports_path(:host_id => host), :title => _("Browse host config management reports"), :class => 'btn btn-default'), :priority => 300 } if host.reports.any?),
-      { :button => link_to(_("New UI"), host_details_page_path(host.name), :title => _("Switch to the new host details UI"), :class => 'btn btn-default'), :priority => 50 },
+      { :button => link_to(_("New UI"), "#{host_details_page_path(host.name)}#/Overview", :title => _("Switch to the new host details UI"), :class => 'btn btn-default'), :priority => 50 },
```
Comment (ezr-ondrej) [inline app/helpers/host_description_helper.rb:139]: Why are we forcing the `Overview` tab here?

**Review by ezr-ondrej — COMMENTED** [review]
Few nitpicks, please answer what you think, but I believe we can merge as is.

Tho I've noticed, the tabs are not clickable for some reason (not caused by this PR tho)

**Comment by amirfefer** [comment]
@ezr-ondrej, Thanks.
To be honest, I don't like this approach, due to integrations errors I added more and more workarounds, which I prefer not to merge.

I have another approach,  more straightforward and cleaner, I'll open a new PR for it so we can compare the two.

**Comment by ezr-ondrej** [comment]
> To be honest, I don't like this approach, due to integrations errors I added more and more workarounds, which I prefer not to merge.

Ok :) will wait for the new PR ;)

**Comment by amirfefer** [comment]
Following https://github.com/theforeman/foreman/pull/8780#issuecomment-952301868 closing this
